### DISCUSSION TITLE
Adding trailer metadata in gRPC server error cause the response to fail with UNKNOWN status

### DISCUSSION THREAD

**Discussion by jdussouillez** [quarkusio/quarkus#31110] [Q&A] [answered]
I don't know if it's a bug or not, I'm not a gRPC expert.

When mapping an exception to a gRPC `StatusException` with a custom status and metadata then the response fails with `UNKNOWN` status.

```java
public StatusException toGrpc(final CustomException ex) {
    var trailers = new Metadata();
    trailers.put(Metadata.Key.of("foo", Metadata.ASCII_STRING_MARSHALLER), "bar");
    return Status.fromCode(Code.INTERNAL)
        .withDescription(ex.getMessage())
        .withCause(ex)
        .asException(trailers);
}
```

For example, this ends with status `UNKNOWN` and no trailer metadata. If I comment the line `trailers.put(...)` then it works fine (the status is `INTERNAL`).
Is this a bug or do I miss something here?

I found this while trying to resolve this issue : https://stackoverflow.com/questions/75412735/quarkus-grpc-exceptionhandlerprovider-does-not-close-response-message-with-the-c (see the end of the question: "Another weird thing is [...]")

Repo to reproduce : https://github.com/jdussouillez/quarkus-grpc-exception-handler

[a comment by an automated account is not shown]

**Comment by jdussouillez** [comment]
Ok so it seems to work using the Quarkus gRPC client (status is `INTERNAL` and trailer is present). So maybe it's a Postman issue. Closing this.

**Comment by alesj** [comment, marked as answer]
This is a known issue ... 

What's the client-server combination you're using?
I see this with using old Netty gRPC client, and our new Vert.x gRPC server:
* https://github.com/alesj/quarkus_cl130/tree/trailersvx1

And this is already fixed in Vert.x upstream ... so it should be fixed once Quarkus gets this new Vert.x version ...

**Reply by cescoffier** [reply]
I just opened the PR bumping Vert.x

**Reply by jdussouillez** [reply]
Oh OK thanks, I'll wait for the next Quarkus release then.

----

I'm using the Quarkus v2.16.2 gRPC server ([the new one based on Vert.X using "use-separate-server" flag](https://quarkus.io/guides/grpc-service-implementation#quarkus-grpc-config-group-config-grpc-server-configuration_quarkus.grpc.server.use-separate-server))

Same thing for the gRPC client : Quarkus v2.16.0 [based on Vert.X using "use-quarkus-grpc-client" flag](https://quarkus.io/guides/grpc-service-consumption#quarkus-grpc-config-group-config-grpc-client-configuration_quarkus.grpc.clients.-client-name-.use-quarkus-grpc-client)

The status/trailers bug happens when calling the gRPC server from Postman.

**Comment by alesj** [comment]
Yeah, I think new Vert.x gRPC server has this trailers issue ... 
@cescoffier we must make sure we also change the test that initiated all this -- to actually have trailers :-)

**Comment by hennihaus** [comment]
Is this issue already fixed in a new Quarkus release?

**Reply by cescoffier** [reply]
Yes, I think so.

**Comment by VladimirYushkevich** [comment]
I came across this [question](https://stackoverflow.com/questions/75412735/quarkus-grpc-exceptionhandlerprovider-does-not-close-response-message-with-the-c) and found that the `handleException` method from the [doc](https://quarkus.io/guides/grpc-service-consumption#custom-exception-handling) is not being called for me either. I'm using `quarkus:3.14.2`. My question is more about curiosity what is a purpose of this method in custom ExceptionHandlerProvider.

**Reply by cescoffier** [reply]
` return new StatusRuntimeException(Status.ABORTED.withDescription(he.getName()));` should be calling the method. The documentation needs to be amended.

\CC @alesj

**Comment by alesj** [comment]
Ah, OK ... I see what you mean.

The EHProvider impl (in docs case HelloExceptionHandlerProvider) is actually responsible for 2 things
(a) creating ServerCall.Listener wrapper ... aka ExceptionHandler ... in docs case HelloExceptionHandler
(b) providing EHP::transform()

Unfortunately it turns out that the current impl couldn't handle all exceptions with (a),
hence (b) was introduced - to be used in those few cases (a) doesn't kick-in, 
so to be as much as possible consistent with user's custom exception handling.

My guess is that in your case (b) is called.

I guess we need to document this a bit better, explaining / mentioning ^^ ...
